google-cloud-logging/src/main/java/com/google/cloud/logging/MonitoredResourceUtil.java

```diff
@@ -0,0 +1,208 @@
+    // between "gae_app_flex", "gae_app_standard" to support zone id, instance name logging on flex VMs.
+    // Hence, "gae_app_flex", "gae_app_standard" are trimmed to "gae_app"
+    String resourceName = resourceType.startsWith("gae_app") ? "gae_app" : resourceType;
+    MonitoredResource.Builder builder =
+        MonitoredResource.newBuilder(resourceName).addLabel(Label.ProjectId.getKey(), projectId);
+    Label[] resourceLabels = resourceTypeWithLabels.get(resourceType);
+    if (resourceLabels != null) {
+      for (Label label : resourceLabels) {
+        String value = getValue(label);
+        if (value != null) {
+          builder.addLabel(label.getKey(), value);
+        }
+      }
+    }
+    return builder.build();
+  }
+
+  /**
+   * Returns custom log entry enhancers (if available) for resource type.
+   * @return custom long entry enhancers
+   */
+  public static List<LoggingEnhancer> getResourceEnhancers() {
+    Resource resourceType = getAutoDetectedResourceType();
+    return getEnhancers(resourceType);
+  }
+
+  private static String getValue(Label label) {
+    String value;
+    switch (label) {
+      case AppId:
+        value = ServiceOptions.getAppEngineAppId();
+        break;
+      case ClusterName:
+        value = MetadataConfig.getClusterName();
+        break;
+      case InstanceId:
+        value = MetadataConfig.getInstanceId();
+        break;
+      case InstanceName:
+        value = getAppEngineInstanceName();
+        break;
+      case ModuleId:
+        value = getAppEngineModuleId();
+        break;
+      case VersionId:
+        value = getAppEngineVersionId();
+        break;
+      case Zone:
+        value = MetadataConfig.getZone();
+        break;
+      default:
+        value = null;
+        break;
+    }
+    return value;
+  }
+
+  /* Detect monitored Resource type using environment variables, else return global as default. */
+  private static Resource getAutoDetectedResourceType() {
+    if (System.getenv("GAE_INSTANCE") != null) {
+      return Resource.GaeAppFlex;
+    }
+    if (System.getenv("KUBERNETES_SERVICE_HOST") != null) {
+      return Resource.Container;
+    }
+    if (ServiceOptions.getAppEngineAppId() != null) {
+      return Resource.GaeAppStandard;
+    }
+    if (MetadataConfig.getInstanceId() != null) {
+      return Resource.GceInstance;
+    }
+    // default Resource type
+    return Resource.Global;
+  }
+
+  private static String getAppEngineModuleId() {
+    return System.getenv("GAE_SERVICE");
+  }
+
+  private static String getAppEngineVersionId() {
+    return System.getenv("GAE_VERSION");
+  }
+
+  private static String getAppEngineInstanceName() {
+    return System.getenv("GAE_INSTANCE");
+  }
+
+  private static List<LoggingEnhancer> getEnhancers(Resource resourceType) {
+    List<LoggingEnhancer> enhancers;
+    switch (resourceType) {
+      case GaeAppFlex:
+        enhancers = new ArrayList<>();
+        enhancers.add(new TraceLoggingEnhancer());
```

Comment:
GaeAppStandard also has traceId available in the requests, although only GaeAppFlex will have the code to call the static setter